File: 0895-shortest-path-to-get-all-keys/0895-shortest-path-to-get-all-keys.py
from collections import deque
from typing import List

class Solution:
    def shortestPathAllKeys(self, grid: List[str]) -> int:
        R, C = len(grid), len(grid[0])
        total_keys = 0
        queue = deque()
        visited = [[[False] * (1 << 6) for _ in range(C)] for _ in range(R)]

        # Find starting point and count keys
        for row in range(R):
            for col in range(C):
                ch = grid[row][col]
                if ch == '@':
                    queue.append((row, col, 0, 0))  # row, col, steps, keys bitmask
                    visited[row][col][0] = True
                elif 'a' <= ch <= 'f':
                    total_keys += 1

        directions = [(0, 1), (1, 0), (-1, 0), (0, -1)]
        final_key_mask = (1 << total_keys) - 1

        while queue:
            r, c, steps, keys = queue.popleft()

            if keys == final_key_mask:
                return steps

            for dr, dc in directions:
                nr, nc = r + dr, c + dc

                if 0 <= nr < R and 0 <= nc < C:
                    ch = grid[nr][nc]
                    if ch == '#':
                        continue

                    new_keys = keys
                    if 'a' <= ch <= 'f':
                        new_keys |= 1 << (ord(ch) - ord('a'))

                    if 'A' <= ch <= 'F' and not (keys & (1 << (ord(ch) - ord('A')))):
                        continue  # locked door without key

                    if not visited[nr][nc][new_keys]:
                        visited[nr][nc][new_keys] = True
                        queue.append((nr, nc, steps + 1, new_keys))

        return -1

File: 0895-shortest-path-to-get-all-keys/0895-shortest-path-to-get-all-keys.java
class Solution {
    public int shortestPathAllKeys(String[] grid) {
        int R= grid.length;
        int C= grid[0].length();
        boolean visited[][][] = new boolean[R][C][1<<6];
        Queue<int[]> queue = new LinkedList<>();
        int keys=0;
        for(int row=0; row<R; row++){
            for(int col=0;col<C;col++){
                char ch=grid[row].charAt(col);
                if(ch=='@'){
                    queue.add(new int[]{row,col,0,0});
                    visited[row][col][0] = true;
                }
                else if(ch>='a' && ch<='z'){
                    keys++;
                }
            }
        }
        int[][] diff = {{0,1},{1,0},{-1,0},{0,-1}};
        while(!queue.isEmpty()){
            int cell[] = queue.poll();
            int row = cell[0],col=cell[1],moves=cell[2],flag = cell[3];
            if (flag == ((1<<keys)-1)){
                return moves;
            }
            for(int i=0;i<4;i++){
                int ar=row+diff[i][0];
                int ac=col+diff[i][1];
                if(ar>=0 && ar< R && ac>=0 && ac<C){
                    char ch = grid[ar].charAt(ac);
                    if(ch=='#'){
                        continue;
                    }
                    if(ch>='A' && ch<='F'){
                        if ((flag & (1<<(ch-'A')))==0){
                            continue;
                        }
                    }
                    int new_flag = flag;
                    if(ch >='a' && ch<='f'){
                        new_flag = new_flag | (1<<(ch-'a'));
                    }
                    if(!visited[ar][ac][new_flag]){
                        queue.add(new int[]{ar,ac,moves+1,new_flag});
                        visited[ar][ac][new_flag]=true;
                    }
                }
            }
        }
        return -1;
    }
}